components/nimbus/src/behavior.rs

```diff
@@ -0,0 +1,415 @@
+    fn value(&self) -> String {
+        match *self {
+            IntervalName::Example1 => "example1".to_string(),
+            IntervalName::Example2 => "example2".to_string(),
+        }
+    }
+}
+
+impl PartialEq for IntervalName {
+    fn eq(&self, other: &Self) -> bool {
+        self.value() == other.value()
+    }
+}
+impl Eq for IntervalName {}
+
+impl Hash for IntervalName {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.value().as_bytes().hash(state);
+    }
+}
+
+#[derive(Clone)]
+struct IntervalConfig {
+    bucket_count: usize,
+    interval: Interval,
+}
+
+impl Default for IntervalConfig {
+    fn default() -> Self {
+        Self::new(7, Interval::Days)
+    }
+}
+
+impl IntervalConfig {
+    fn new(bucket_count: usize, interval: Interval) -> Self {
+        Self {
+            bucket_count,
+            interval,
+        }
+    }
+}
+
+pub struct IntervalData {
+    buckets: VecDeque<u64>,
+    bucket_count: usize,
+    starting_instant: DateTime<Utc>,
+}
+
+impl Default for IntervalData {
+    fn default() -> Self {
+        Self::new(1)
+    }
+}
+
+impl IntervalData {
+    fn new(bucket_count: usize) -> Self {
+        let mut data = Self {
+            buckets: VecDeque::new(),
+            bucket_count,
+            starting_instant: Utc::now(),
+        };
+        data.buckets.push_front(0);
+        data
+    }
+
+    pub fn from(
+        buckets: VecDeque<u64>,
+        bucket_count: usize,
+        starting_instant: DateTime<Utc>,
+    ) -> Self {
+        Self {
+            buckets,
+            bucket_count,
+            starting_instant,
+        }
+    }
+
+    fn increment(&mut self) -> Result<()> {
+        match self.buckets.front_mut() {
+            Some(x) => *x += 1,
+            None => {
+                return Err(NimbusError::BehaviorError(BehaviorError::InvalidState(
+                    "Interval buckets cannot be empty".to_string(),
+                )))
+            }
+        };
+        Ok(())
+    }
+
+    fn rotate(&mut self, num_rotations: u64) -> Result<()> {
+        for _ in 1..=num_rotations {
+            self.buckets.push_front(0)
+        }
+        if self.buckets.len() > self.bucket_count {
+            self.buckets.drain(self.bucket_count..);
+        }
+        Ok(())
+    }
+}
+
+fn date_diff(a: DateTime<Utc>, b: DateTime<Utc>) -> u32 {
+    (a.date() - b.date()).num_days().try_into().unwrap()
+}
+
+struct SingleIntervalCounter {
+    data: IntervalData,
+    config: IntervalConfig,
+}
+
+impl SingleIntervalCounter {
+    pub fn new(config: IntervalConfig) -> Self {
+        Self {
+            data: IntervalData::new(config.bucket_count),
+            config,
+        }
+    }
+
+    pub fn from(data: IntervalData, config: IntervalConfig) -> Self {
+        Self { data, config }
+    }
+
+    pub fn increment(&mut self) -> Result<()> {
+        self.data.increment()
+    }
+
+    pub fn maybe_advance(&mut self, now: DateTime<Utc>) -> Result<()> {
```

Comment:
Shouldn't this update the `starting_instant` so we know the last time we rotated this data was?

Reply:
Ah I didn't think about that, yep going to update that.